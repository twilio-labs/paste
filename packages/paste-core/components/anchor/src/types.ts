export type AnchorTargets = '_self' | '_blank' | '_parent' | '_top';
export type AnchorTabIndexes = 0 | -1;

export interface AnchorProps {
  id?: never;
  className?: never;
  href: string;
  tabIndex?: AnchorTabIndexes;
  target?: AnchorTargets;
  rel?: string;

  children: NonNullable<React.ReactNode>;

  onClick?(event: React.MouseEvent<HTMLElement>): void;
  onFocus?(event: React.FocusEvent<HTMLElement>): void;
  onBlur?(event: React.FocusEvent<HTMLElement>): void;
}
