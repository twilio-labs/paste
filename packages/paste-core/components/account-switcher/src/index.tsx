export * from "./AccountSwitcher";
export * from "./AccountSwitcherGroup";
export * from "./AccountSwitcherSeparator";
export * from "./AccountSwitcherItem";
export * from "./AccountSwitcherItemRadio";
export * from "./AccountSwitcherBadge";
export * from "./useAccountSwitcher";
