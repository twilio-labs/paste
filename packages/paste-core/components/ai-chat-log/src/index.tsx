export { AIChatMessage } from "./AIChatMessage";
export type { AIChatMessageProps } from "./AIChatMessage";
export { AIChatMessageActionGroup } from "./AIChatMessageActionGroup";
export type { AIChatMessageActionGroupProps } from "./AIChatMessageActionGroup";
export { AIChatMessageBody } from "./AIChatMessageBody";
export type { AIChatMessageBodyProps } from "./AIChatMessageBody";
export { AIChatMessageAuthor } from "./AIChatMessageAuthor";
export type { AIChatMessageAuthorProps } from "./AIChatMessageAuthor";
export { AIChatMessageActionCard } from "./AIChatMessageActionCard";
export type { AIChatMessageActionCardProps } from "./AIChatMessageActionCard";
export { AIChatMessageLoading } from "./AIChatMessageLoading";
export type { AIChatMessageLoadingProps } from "./AIChatMessageLoading";

export { AIChatLog } from "./AIChatLog";
export type { AIChatLogProps } from "./AIChatLog";
export { useAIChatLogger } from "./useAIChatLogger";
export type { UseAIChatLogger, AIPartialIDChat, AIChat } from "./useAIChatLogger";
export { AIChatLogger } from "./AIChatLogger";
export type { AIChatLoggerProps } from "./AIChatLogger";

export type { AIMessageVariants } from "./AIMessageContext";
