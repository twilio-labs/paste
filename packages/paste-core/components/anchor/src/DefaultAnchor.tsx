import * as React from 'react';
import {Box} from '@twilio-paste/box';
import {AnchorProps, AnchorPropTypes} from './types';

const DefaultAnchor = React.forwardRef<HTMLAnchorElement, AnchorProps>((props, ref) => (
  <Box
    {...props}
    as="a"
    color="colorTextLink"
    fontSize="inherit"
    fontWeight="inherit"
    lineHeight="inherit"
    outline="none"
    ref={ref}
    textDecoration="underline"
    _active={{
      color: 'colorTextLinkStronger',
      textDecoration: 'none',
    }}
    _focus={{
      boxShadow: 'shadowFocus',
      color: 'colorTextLinkStronger',
      textDecoration: 'none',
    }}
    _hover={{
      color: 'colorTextLinkStronger',
      textDecoration: 'none',
    }}
  >
    {props.children}
  </Box>
));

DefaultAnchor.displayName = 'DefaultAnchor';

if (process.env.NODE_ENV === 'development') {
  DefaultAnchor.propTypes = AnchorPropTypes;
}

export {DefaultAnchor};
