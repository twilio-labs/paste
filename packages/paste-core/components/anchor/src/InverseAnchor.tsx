import * as React from 'react';
import {Box} from '@twilio-paste/box';
import {AnchorProps, AnchorPropTypes} from './types';

const InverseAnchor = React.forwardRef<HTMLAnchorElement, AnchorProps>((props, ref) => (
  <Box
    {...props}
    as="a"
    color="colorTextInverse"
    fontSize="inherit"
    fontWeight="inherit"
    lineHeight="inherit"
    outline="none"
    ref={ref}
    textDecoration="underline"
    _active={{
      color: 'colorTextInverse',
      textDecoration: 'none',
    }}
    _focus={{
      boxShadow: 'shadowFocusInverse',
      color: 'colorTextInverse',
      textDecoration: 'none',
    }}
    _hover={{
      color: 'colorTextInverse',
      textDecoration: 'none',
    }}
  >
    {props.children}
  </Box>
));

InverseAnchor.displayName = 'InverseAnchor';

if (process.env.NODE_ENV === 'development') {
  InverseAnchor.propTypes = AnchorPropTypes;
}

export {InverseAnchor};
