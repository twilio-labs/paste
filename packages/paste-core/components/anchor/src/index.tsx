import * as React from 'react';
import * as PropTypes from 'prop-types';
import {compose, layout, space} from 'styled-system';
import styled from '@emotion/styled';
import css from '@styled-system/css';
import {LayoutProps, SpaceProps} from '@twilio-paste/style-props';

export type AnchorTargets = '_self' | '_blank' | '_parent' | '_top';
export type AnchorTabIndexes = 0 | -1;

export interface AnchorProps extends React.AnchorHTMLAttributes<HTMLAnchorElement>, LayoutProps, SpaceProps {
  children: NonNullable<React.ReactNode>;
  href: string;
  rel?: string;
  tabIndex?: AnchorTabIndexes;
  target?: AnchorTargets;
  id?: never;
  className?: never;
}

const EXTERNAL_URL_REGEX = /^(https?:)[^\s]*$/;
const EXTERNAL_TARGET_DEFAULT = '_blank';
const EXTERNAL_REL_DEFAULT = 'noreferrer noopener';
const isExternalUrl = (url: string): boolean => EXTERNAL_URL_REGEX.test(url);

export const secureExternalLink = (href: string): {} | undefined => {
  if (!isExternalUrl(href)) return undefined;
  return {
    rel: EXTERNAL_REL_DEFAULT,
    target: EXTERNAL_TARGET_DEFAULT,
  };
};

// eslint-disable-next-line emotion/syntax-preference
const StyledAnchor = styled.a(
  css({
    color: 'colorTextLink',
    textDecoration: 'underline',
    outline: 'none',

    '&:hover': {
      color: 'colorTextLinkDarker',
      textDecoration: 'none',
    },

    '&:focus, &:active': {
      boxShadow: 'shadowFocus',
      color: 'colorTextLinkDarker',
      textDecoration: 'none',
    },
  }),
  compose(
    space,
    layout
  )
);

const Anchor: React.FC<AnchorProps> = props => (
  <StyledAnchor {...secureExternalLink(props.href)} {...props}>
    {props.children}
  </StyledAnchor>
);
Anchor.displayName = 'Anchor';

if (process.env.NODE_ENV === 'development') {
  Anchor.propTypes = {
    children: PropTypes.node.isRequired,
    href: PropTypes.string.isRequired,
    rel: PropTypes.string,
    tabIndex: PropTypes.oneOf([0, -1]),
    target: PropTypes.oneOf(['_self', '_blank', '_parent', '_top']),
  };
}
export {Anchor};
