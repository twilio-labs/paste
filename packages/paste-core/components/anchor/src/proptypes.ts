import * as PropTypes from 'prop-types';

export const AnchorPropTypes = {
  children: PropTypes.node.isRequired,
  href: PropTypes.string.isRequired,
  rel: PropTypes.string,
  showExternal: PropTypes.bool,
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  tabIndex: PropTypes.oneOf([0, -1]) as any,
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  target: PropTypes.oneOf(['_self', '_blank', '_parent', '_top']) as any,
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  variant: PropTypes.oneOf(['default', 'inverse']) as any,
  /** element identifier for customization */
  element: PropTypes.string,
};
