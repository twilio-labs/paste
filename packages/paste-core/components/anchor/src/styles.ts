import styled from '@emotion/styled';
import {themeGet} from '@styled-system/theme-get';

// Link
export const StyledLink = styled.a`
  color: ${themeGet('textColors.colorTextLink')};
  text-decoration: none;
  outline: none;

  &:hover {
    color: ${themeGet('textColors.colorTextLinkHover')};
    text-decoration: underline;
  }

  &:focus,
  &:active {
    box-shadow: ${themeGet('shadows.shadowFocus')};
    text-decoration: underline;
  }

  &:focus {
    color: ${themeGet('textColors.colorTextLinkFocus')};
  }

  &:active {
    color: ${themeGet('textColors.colorTextLinkActive')};
  }
`;
