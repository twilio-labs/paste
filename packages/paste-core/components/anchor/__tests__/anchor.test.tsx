import * as React from 'react';
import renderer from 'react-test-renderer';
import {Theme} from '@twilio-paste/theme';
import {Anchor} from '../src';

describe('Anchor', () => {
  it('it should render an anchor', (): void => {
    const tree = renderer
      .create(
        <Theme.Provider theme="console">
          <Anchor href="/">This is an anchor</Anchor>
        </Theme.Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('it should render an external anchor', (): void => {
    const tree = renderer
      .create(
        <Theme.Provider theme="console">
          <Anchor href="https://twilio.com">
            This is an anchor that links to Twilio.com with an external target and rel
          </Anchor>
        </Theme.Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('it should render an external anchor with overwritten target and rel', (): void => {
    const tree = renderer
      .create(
        <Theme.Provider theme="console">
          <Anchor href="https://twilio.com" target="_self" rel="noopener">
            This is an anchor that links to Twilio.com with the target and rel overwritten
          </Anchor>
        </Theme.Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
