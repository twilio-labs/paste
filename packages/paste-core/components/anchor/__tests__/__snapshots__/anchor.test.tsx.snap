// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Anchor it should render an anchor 1`] = `
.emotion-2 {
  font-family: 'Whitney SSm A','Whitney SSm B','Helvetica Neue',Helvetica,Arial,sans-serif;
  font-size: 0.875rem;
  line-height: 1.5rem;
  color: #282a2b;
  font-weight: 400;
}

.emotion-2 *,
.emotion-2 *::after,
.emotion-2 *::before {
  box-sizing: border-box;
}

.emotion-0 {
  color: #005ea6;
  -webkit-text-decoration: underline;
  text-decoration: underline;
  outline: none;
}

.emotion-0:hover {
  color: #001f57;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.emotion-0:focus,
.emotion-0:active {
  box-shadow: 0 0 0 4px rgba(0,117,195,0.5);
  color: #001f57;
  -webkit-text-decoration: none;
  text-decoration: none;
}

<div
  className="emotion-2"
>
  <a
    className="emotion-0 emotion-1"
    href="/"
  >
    This is an anchor
  </a>
</div>
`;

exports[`Anchor it should render an external anchor 1`] = `
.emotion-2 {
  font-family: 'Whitney SSm A','Whitney SSm B','Helvetica Neue',Helvetica,Arial,sans-serif;
  font-size: 0.875rem;
  line-height: 1.5rem;
  color: #282a2b;
  font-weight: 400;
}

.emotion-2 *,
.emotion-2 *::after,
.emotion-2 *::before {
  box-sizing: border-box;
}

.emotion-0 {
  color: #005ea6;
  -webkit-text-decoration: underline;
  text-decoration: underline;
  outline: none;
}

.emotion-0:hover {
  color: #001f57;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.emotion-0:focus,
.emotion-0:active {
  box-shadow: 0 0 0 4px rgba(0,117,195,0.5);
  color: #001f57;
  -webkit-text-decoration: none;
  text-decoration: none;
}

<div
  className="emotion-2"
>
  <a
    className="emotion-0 emotion-1"
    href="https://twilio.com"
    rel="noreferrer noopener"
    target="_blank"
  >
    This is an anchor that links to Twilio.com with an external target and rel
  </a>
</div>
`;

exports[`Anchor it should render an external anchor with overwritten target and rel 1`] = `
.emotion-2 {
  font-family: 'Whitney SSm A','Whitney SSm B','Helvetica Neue',Helvetica,Arial,sans-serif;
  font-size: 0.875rem;
  line-height: 1.5rem;
  color: #282a2b;
  font-weight: 400;
}

.emotion-2 *,
.emotion-2 *::after,
.emotion-2 *::before {
  box-sizing: border-box;
}

.emotion-0 {
  color: #005ea6;
  -webkit-text-decoration: underline;
  text-decoration: underline;
  outline: none;
}

.emotion-0:hover {
  color: #001f57;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.emotion-0:focus,
.emotion-0:active {
  box-shadow: 0 0 0 4px rgba(0,117,195,0.5);
  color: #001f57;
  -webkit-text-decoration: none;
  text-decoration: none;
}

<div
  className="emotion-2"
>
  <a
    className="emotion-0 emotion-1"
    href="https://twilio.com"
    rel="noopener"
    target="_blank"
  >
    This is an anchor that links to Twilio.com with the target and rel overwritten
  </a>
</div>
`;
